File: cpp/1090. 受标签影响的最大值.cpp
class Solution {
public:
    int largestValsFromLabels(vector<int>& values, vector<int>& labels, int numWanted, int useLimit) {
        int n=values.size();
        vector<int> id(n);
        iota(id.begin(),id.end(),0);
        sort(id.begin(),id.end(),[&](int x,int y){
            return values[x]>values[y];
        });
        unordered_map<int,int> used;
        int sum=0;
        for(int i=0;i<n;i++){
            int value=values[id[i]];
            if(used[labels[id[i]]]==useLimit){
                continue;
            }else if(numWanted==0){
                break;
            }else{
                sum+=value;
                numWanted--;
                used[labels[id[i]]]++;
            }
        }
        return sum;

    }
};

File: golang/1090. 受标签影响的最大值.go
func largestValsFromLabels(values []int, labels []int, numWanted int, useLimit int) int {
	n := len(values)
	ids := make([]int, n)
	for i := 0; i < n; i++ {
		ids[i] = i
	}
	sort.Slice(ids, func(i, j int) bool {
		return values[ids[i]] > values[ids[j]]
	})
	ans, choose := 0, 0
	cnt := make(map[int]int)
	for i := 0; i < n; i++ {
		label := labels[ids[i]]
		if cnt[label] == useLimit {
			continue
		}
		choose++
		ans += values[ids[i]]
		cnt[label]++
		if choose == numWanted {
			break
		}
	}
	return ans
}